src/coreclr/vm/gchelpers.cpp

```diff
@@ -183,6 +183,76 @@ inline void CheckObjectSize(size_t alloc_size)
     }
 }
 
+// TODO: how to define default sampling distribution mean?
+#define g_SamplingDistributionMean (100 * 1024)
+
+// TODO: what are the call sites other than the fast path helpers?
+// if there is not,
+//      isSampled =
+//          (pEEAllocContext->alloc_sampling != nullptr) &&
+//          (pEEAllocContext->alloc_sampling != pAllocContext->alloc_limit);
+//
+inline Object* Alloc(ee_alloc_context* pEEAllocContext, size_t size, GC_ALLOC_FLAGS flags)
+{
+    CONTRACTL {
+        THROWS;
+        GC_TRIGGERS;
+        MODE_COOPERATIVE; // returns an objref without pinning it => cooperative
+    } CONTRACTL_END;
+
+    Object* retVal = nullptr;
+    gc_alloc_context* pAllocContext = &pEEAllocContext->gc_alloc_context;
+
+    // TODO: isSampled could be an out parameter used by PublishObjectAndNotify()
+    //       where the event would be emitted
+    bool isSampled = false;
+    size_t samplingBudget = (size_t)(pAllocContext->alloc_limit - pEEAllocContext->alloc_sampling);
+    size_t availableSpace = (size_t)(pAllocContext->alloc_limit - pAllocContext->alloc_ptr);
+    if (flags & GC_ALLOC_USER_OLD_HEAP)
+    {
+        // TODO: if sampling is on, decide if this object should be sampled
+        // the provided allocation context fields are not used for LOH/POH allocations
+        // (only its alloc_bytes_uoh field will be updated)
+        // so get a random size in the distribution and if it is less than the size of the object
+        // then this object should be sampled
+    }
+    else
+    {
+        // check if this allocation overlaps the SOH sampling point:
+        //    if the sampling threashold was outside of the allocation context,
+        //    then
+        //      alloc_sampling was set to alloc_limit
+        //      or the allocation context was just initialized to nullptr for the first allocation
+        //    else
+        //      use the alloc_sampling to decide if this allocation should be sampled
+        // see the comment at the beginning of the function for the simplified check
+        isSampled =
+            (pEEAllocContext->alloc_sampling != nullptr) &&  // nullptr first time the allocation context is used
```

Comment:
I think I get your reasoning: if (alloc_sampling == alloc_limit) then we should not sample
Am I right?

In my mind, I thought it could be possible to sample on the last byte of the allocation context (i.e. sampling == limit)


But do we need to check (size > samplingBudget)?
As I wrote in the function header comment, it should not be needed if this function is called only by the fast path helpers